File: src/main/java/com/almasb/zeph/ui/EquipmentView.java
package com.almasb.zeph.ui;

import com.almasb.fxgl.texture.Texture;
import com.almasb.zeph.character.CharacterEntity;
import com.almasb.zeph.character.EquipPlace;
import com.almasb.zeph.item.Item;
import javafx.scene.Cursor;
import javafx.scene.Group;
import javafx.scene.Parent;
import javafx.scene.control.Tooltip;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import javafx.scene.shape.Circle;
import javafx.scene.shape.Rectangle;
import javafx.scene.shape.Shape;
import javafx.scene.text.Font;
import javafx.scene.text.Text;

import java.util.HashMap;
import java.util.Map;

import static com.almasb.fxgl.dsl.FXGL.texture;

public final class EquipmentView extends Parent {

    private Map<EquipPlace, Group> groups = new HashMap<>();

    private CharacterEntity player;

    public EquipmentView(CharacterEntity player) {
        this.player = player;

        groups.put(EquipPlace.HELM, createGroup(88, 60));
        groups.put(EquipPlace.BODY, createGroup(88, 105));
        groups.put(EquipPlace.SHOES, createGroup(88, 150));
        groups.put(EquipPlace.LEFT_HAND, createGroup(133, 105));
        groups.put(EquipPlace.RIGHT_HAND, createGroup(43, 105));

        for (EquipPlace place : EquipPlace.values()) {
            setItem(place, player.getPlayerComponent().getEquip(place));
            player.getPlayerComponent().equipProperty(place).addListener((obs, old, newItem) -> {
                setItem(place, newItem);
            });
        }

        Texture background = texture("ui/inventory_left.png");

        Rectangle border = new Rectangle(200, 240);
        border.setStrokeWidth(2);
        border.setArcWidth(10);
        border.setArcHeight(10);
        border.setFill(Color.rgb(25, 25, 25, 0.8));
        border.setStroke(Color.WHITE);

        getChildren().addAll(background);
        getChildren().addAll(groups.values());
    }

    private Group createGroup(int x, int y) {
        Group group = new Group();
        group.setTranslateX(x);
        group.setTranslateY(y);
        Tooltip tooltip = new Tooltip();

        Text text = new Text();
        text.setFont(Font.font(20));
        text.setFill(Color.WHITE);
        text.setWrappingWidth(200);

        tooltip.setGraphic(text);
        Tooltip.install(group, tooltip);

        group.setUserData(text);
        return group;
    }

    private void setItem(EquipPlace place, Item item) {
        Texture view = texture(item.getDescription().getTextureName());
        view.setOnMouseClicked(event -> player.getPlayerComponent().unEquipItem(place));
        view.setCursor(Cursor.HAND);

        Group group = groups.get(place);
        group.getChildren().setAll(view);

        ((Text)group.getUserData()).textProperty().bind(item.getDynamicDescription());
    }
}

File: src/main/kotlin/com/almasb/zeph/ui/EquipmentView.kt
package com.almasb.zeph.ui

import com.almasb.fxgl.dsl.FXGL
import com.almasb.fxgl.dsl.texture
import com.almasb.zeph.character.CharacterEntity
import com.almasb.zeph.character.EquipPlace
import com.almasb.zeph.character.EquipPlace.*
import com.almasb.zeph.item.Item
import javafx.geometry.Point2D
import javafx.scene.Cursor
import javafx.scene.Parent

/**
 *
 * @author Almas Baimagambetov (almaslvl@gmail.com)
 */
class EquipmentView(private val player: CharacterEntity) : Parent() {

    companion object {
        private val equipViewLocations = mapOf(
                HELM to Point2D(88.0, 60.0),
                BODY to Point2D(88.0, 105.0),
                SHOES to Point2D(88.0, 150.0),
                LEFT_HAND to Point2D(133.0, 105.0),
                RIGHT_HAND to Point2D(43.0, 105.0)
        )
    }

    init {
        children.addAll(
                texture("ui/inventory_left.png")
        )

        for (place in EquipPlace.values()) {
            val p = equipViewLocations[place]!!
            val equipView = EquipmentItemView(place, player.playerComponent!!.getEquip(place), p.x, p.y)

            player.playerComponent!!.equipProperty(place).addListener { _, _, newItem: Item ->
                equipView.item = newItem
                equipView.updateView()
            }

            children += equipView
        }
    }

    private inner class EquipmentItemView(val place: EquipPlace, var item: Item, x: Double, y: Double) : Parent() {

        init {
            relocate(x, y)

            isPickOnBounds = true
            cursor = Cursor.HAND

            updateView()

            setOnMouseClicked {
                player.playerComponent!!.unEquipItem(place)
            }

            setOnTooltipHover {
                it.setItem(item)
            }
        }

        fun updateView() {
            val itemView = FXGL.texture(item.description.textureName)

            children.setAll(itemView)
        }
    }
}
